test/k8sdeployment/connection_test.go

```diff
@@ -0,0 +1,43 @@
+package k8sdeployment
+
+import (
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/coredns/ci/test/kubernetes"
+)
+
+// This test restarts the kube-APIserver and checks if CoreDNS connection to the API
+// is valid and there are no failures.
+// This test is to catch bugs/errors such as the one reported in https://github.com/coredns/coredns/issues/2464
+func TestConnectionAfterAPIRestart(t *testing.T) {
+	// Apply manifests via coredns/deployment deployment script ...
+	path := os.Getenv("DEPLOYMENTPATH")
+	cmd := exec.Command("sh", "-c", "./deploy.sh -s -i 10.96.0.10 -r 10.96.0.0/8 -r 172.17.0.0/16 | kubectl apply -f -")
+	cmd.Dir = path + "/kubernetes"
+	cmdout, err := cmd.CombinedOutput()
+	if err != nil {
+		t.Fatalf("deployment script failed: %s\nerr: %s", string(cmdout), err)
+	}
+
+	// Restart the Kubernetes APIserver and wait for it to come back up.
+	dockerCmd, err := exec.Command("sh", "-c", "docker restart $(docker ps --no-trunc | grep 'kube-apiserver' | awk '{ print $1; }') > /dev/null").CombinedOutput()
+	if err != nil {
+		t.Fatalf("docker container restart failed: %s\nerr: %s", string(dockerCmd), err)
+	}
+	time.Sleep(15 * time.Second)
```

Comment:
OK - if total duration of the deployment test becomes an issue we should revisit this.  The crash should be immediate, so we should be able to detect the occurrence in a lot less than 15 seconds.  Can this be lowered at all without flaking?

Reply:
There seems to be a delay from when the apiserver is restarted to getting the output for the number of restarts. I tried the sleep to 10secs, but the test flaked and passed. 15 secs is a lot but is reqd for the test to not flake.